File: 백준/Platinum/27451. 마키마씨가 정해주는 오늘 점심의 맛/마키마씨가 정해주는 오늘 점심의 맛.cs
string[] nm = Console.ReadLine()!.Split();
int n = int.Parse(nm[0]), m = int.Parse(nm[1]);

List<int>[] edge = new List<int>[n + 1];
for (int i = 0; i <= n; i++)
{
    edge[i] = new();
}

bool[,,] visited = new bool[n + 1, n + 1, n + 1];

string[] abc = Console.ReadLine()!.Split();
int a = int.Parse(abc[0]), b = int.Parse(abc[1]), c = int.Parse(abc[2]);

visited[a, b, c] = true;

Queue<(int, int, int, string, string, string)> queue = new();
queue.Enqueue((a, b, c, a.ToString(), b.ToString(), c.ToString()));

while (m-- > 0)
{
    string[] path = Console.ReadLine()!.Split();
    int s = int.Parse(path[0]), e = int.Parse(path[1]);
    edge[s].Add(e);
}

int cnt = 0;
while (queue.Count > 0)
{
    int qcnt = queue.Count;
    while (qcnt-- > 0)
    {
        var (pow, den, aki, powPath, denPath, akiPath) = queue.Dequeue();
        if (pow == den && den == aki)
        {
            Console.WriteLine($"{aki} {cnt}");
            Console.WriteLine(powPath);
            Console.WriteLine(denPath);
            Console.WriteLine(akiPath);
            return;
        }

        foreach (int node1 in edge[pow])
        {
            foreach (int node2 in edge[den])
            {
                foreach (int node3 in edge[aki])
                {
                    if (!visited[node1, node2, node3])
                    {
                        visited[node1, node2, node3] = true;
                        queue.Enqueue((node1, node2, node3, $"{powPath} {node1}", $"{denPath} {node2}", $"{akiPath} {node3}"));
                    }
                }
            }
        }
    }
    cnt++;
}

Console.WriteLine(-1);

File: 백준/Platinum/27451. 마키마씨가 정해주는 오늘 점심의 맛/마키마씨가 정해주는 오늘 점심의 맛.rs
use std::{collections::VecDeque, io::{stdin, Read}};

fn main() {
    let mut buffer = String::new();
    stdin().read_to_string(&mut buffer).expect("input panic");
    let mut input = buffer.split_whitespace().flat_map(str::parse::<usize>);

    let n = input.next().unwrap();
    let m = input.next().unwrap();

    let mut edge: Vec<Vec<usize>> = Vec::new();
    let mut visited: Vec<Vec<Vec<bool>>> = Vec::new();
    for i in 0..(n + 1) {
        edge.push(Vec::new());
        visited.push(Vec::new());
        for j in 0..(n + 1) {
            visited[i].push(Vec::new());
            for _ in 0..(n + 1) {
                visited[i][j].push(false);
            }
        }
    }

    let a = input.next().unwrap();
    let b = input.next().unwrap();
    let c = input.next().unwrap();
    visited[a][b][c] = true;
    
    for _ in 0..m {
        let s = input.next().unwrap();
        let e = input.next().unwrap();
        edge[s].push(e);
    }

    let mut deque = VecDeque::from([(a, b, c, a.to_string(), b.to_string(), c.to_string())]);

    let mut cnt = 0;
    let mut re = String::from("-1");
    'search: while deque.len() > 0 {
        let len = deque.len();
        for _ in 0..len {
            let v = deque.pop_front().unwrap();
            if v.0 == v.1 && v.0 == v.2 {
                drop(re);
                re = format!("{} {}\n{}\n{}\n{}", v.0, cnt, v.3, v.4, v.5);
                break 'search;
            }

            for i in edge[v.0].iter() {
                for j in edge[v.1].iter() {
                    for k in edge[v.2].iter() {
                        if !visited[*i][*j][*k] {
                            visited[*i][*j][*k] = true;
                            let mut pow_path = v.3.clone();
                            let mut den_path = v.4.clone();
                            let mut aki_path = v.5.clone();
    
                            let pow_next = format!(" {}", *i);
                            let den_next = format!(" {}", *j);
                            let aki_next = format!(" {}", *k);
    
                            pow_path.push_str(&pow_next);
                            den_path.push_str(&den_next);
                            aki_path.push_str(&aki_next);
                            
                            deque.push_back((*i, *j, *k, pow_path, den_path, aki_path));
                        }
                    }
                }
            }
        }
        cnt += 1;
    }
    println!("{}", re);
}
